ambry-cloud/src/main/java/com/github/ambry/cloud/HelixVcrCluster.java

```diff
@@ -167,4 +169,25 @@ public void close() {
     manager.disconnect();
     helixAdmin.close();
   }
+
+  /**
+   * Register the state model factory corresponding to {@code vcrHelixStateModelName}.
+   * @param vcrHelixStateModelName
+   */
+  private void registerStateModelFactory(String vcrHelixStateModelName) {
+    switch (vcrHelixStateModelName) {
+      case OnlineOfflineSMD.name:
+        manager.getStateMachineEngine()
+            .registerStateModelFactory(vcrHelixStateModelName, new OnlineOfflineHelixVcrStateModelFactory(this));
+        break;
+      case LeaderStandbySMD.name:
+        manager.getStateMachineEngine()
+            .registerStateModelFactory(vcrHelixStateModelName, new LeaderStandbyHelixVcrStateModelFactory(this));
+        break;
+      default:
+        throw new IllegalArgumentException(String.format(
+            "Illegal vcr helix state model name: " + vcrHelixStateModelName + ". Should be one of"
+                + OnlineOfflineSMD.name + ", " + LeaderStandbySMD.name));
+    }
```

Comment:
I suggest adding state model validation into CloudConfig, thus it throws exception at the very beginning (when generating the config) rather than keep initializing other components until participation.
You can refer to `clustermapStateModelDefinition` in ClusterMapConfig.

Reply:
fixed.

Reply:
I have changed to code to pass factory class instead. Let me know what do you think.